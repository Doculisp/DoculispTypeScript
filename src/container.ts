import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};