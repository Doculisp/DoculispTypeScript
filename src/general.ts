import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};