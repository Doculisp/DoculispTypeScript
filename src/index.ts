#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
