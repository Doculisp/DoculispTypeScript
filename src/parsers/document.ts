import { IRegisterable, Valid } from "../types.containers";
import { DocumentMap, DocumentParser, DocumentPart } from "../types.document";
import { Result, fail, ok } from "../types.general";
import * as path from 'node:path';
import { IDocumentSearches, IWhiteSpaceSearches, Searcher } from "../types.textHelpers";
import { HandleValue, IInternals, IParseStepForward, IParser, StepParseResult } from "../types.internal";

function createMap(documentPath: string, parts: DocumentPart[]): DocumentMap {
    return {
        documentPath,
        parts,
    };
}

function id<T>(value: T): T {
    return value;
}


function isStopParsingWhiteSpace(input: string, _line: number, _char: number): Result<'stop' | false> {
    const regex = /\S+/;
    if(regex.test(input)) {
        return ok('stop');
    }
    return ok(false);
}

function doesItStartWithDiscarded(startsWith: RegExp, lineIncrement: (line: number) => number, charIncrement: (char: number, found: string) => number): HandleValue<DocumentPart> {
    return function (input: string, line: number, char: number): StepParseResult<DocumentPart> {
        if(startsWith.test(input)) {
            const found: string = (input.match(startsWith) as any)[0];
            return ok({
                type: 'discard',
                rest: input.slice(found.length),
                line: lineIncrement(line),
                char: charIncrement(char, found),
            });
        }
        return ok(false);
    }
}

function doesItStartWithKeep<T>(startsWith: RegExp, map: (parsed: string) => T, lineIncrement: (line: number) => number, charIncrement: (char: number, found: string) => number): HandleValue<T> {
    return function (input:string, line: number, char: number): StepParseResult<T> {
        if(startsWith.test(input)) {
            const parsed: string = (input.match(startsWith) as any)[0];
            const rest = input.slice(parsed.length);
            return ok({
                type: 'parse result',
                subResult: map(parsed),
                rest,
                line: lineIncrement(line),
                char: charIncrement(char, parsed),
            });
        }
        return ok(false);
    }
}

function isDiscardedWhiteSpace(doesIt: IDocumentSearches, createParser: (...handlers: HandleValue<DocumentPart>[]) => IParser<DocumentPart>): HandleValue<DocumentPart> {
    return function (input: string, line: number, char: number): StepParseResult<DocumentPart> {
        const isWindows = doesItStartWithDiscarded(doesIt.startWithWindowsNewline, l => l + 1, () => 1);
        const isLinux = doesItStartWithDiscarded(doesIt.startWithLinuxNewline, l => l + 1, () => 1);
        const isMac = doesItStartWithDiscarded(doesIt.startWithMacsNewline, l => l + 1, () => 1);
        const isWhiteSpace = doesItStartWithDiscarded(doesIt.startWithWhiteSpace, l => l, (c, f) => c + f.length);

        const whiteSpaceParser = createParser(isWindows, isLinux, isMac, isWhiteSpace, isStopParsingWhiteSpace);
        const parsed = whiteSpaceParser.parse(input, line, char);
        if(parsed.success) {
            const [_, leftovers] = parsed.value;
            if(leftovers.remaining === input) {
                return ok(false);
            } else {
                return ok({
                    type: 'discard',
                    rest: leftovers.remaining,
                    line: leftovers.location.line,
                    char: leftovers.location.char,
                });
            }
        } else {
            return parsed;
        }
    }
}

function documentParse(doesIt: IDocumentSearches, parserBuilder: IInternals): Valid<DocumentParser> {
    
    function isKeptWhiteSpace<T>(_documentPath: string, map: (value: string) => T): HandleValue<T> {
        const it = doesIt as IWhiteSpaceSearches;
        return function (input: string, line: number, char: number): StepParseResult<T> {
            const isWindows = doesItStartWithKeep(it.startWithWindowsNewline, map, l => l + 1, () => 1);
            const isLinux = doesItStartWithKeep(it.startWithLinuxNewline, map, l => l + 1, () => 1);
            const isMac = doesItStartWithKeep(it.startWithMacsNewline, map, l => l + 1, () => 1);
            const isWhiteSpace = doesItStartWithKeep(it.startWithWhiteSpace, map, id, (c, f) => c + f.length);

            const parser = parserBuilder.createParser(isWindows, isLinux, isMac, isWhiteSpace, isStopParsingWhiteSpace)
            const parsed = parser.parse(input, line, char);
            if(parsed.success) {
                const [result, leftover] = parsed.value;
                if(leftover.location.line === line && leftover.location.char === char) {
                    return ok(false);
                }

                let step: IParseStepForward = {
                    rest: leftover.remaining,
                    line: leftover.location.line,
                    char: leftover.location.char,
                }

                if(0 === result.length) {
                    return ok(parserBuilder.buildStepParse(step, {
                        type: 'discard',
                    }));
                }

                if(1 === result.length) {
                    return ok(parserBuilder.buildStepParse(step, {
                        type: 'parse result',
                        subResult: result[0] as T
                    }));
                }

                return ok(parserBuilder.buildStepParse(step, {
                    type: 'parse group result',
                    subResult: result.map(r => { return { type:'keep', keptValue: r }; }),
                }));
            }

            return parsed;
        }
    }

    function isMultiline(documentPath: string) : HandleValue<DocumentPart> {
        return function(toParse: string, startingLine: number, startingChar: number): StepParseResult<DocumentPart> {
            let opened: boolean = false;

            function tryParseMultiline(input: string, line: number, char: number): StepParseResult<string> {
                if(doesIt.startWithMultilineMarker.test(input)) {
                    opened = !opened;

                    const parsed: string = (input.match(doesIt.startWithMultilineMarker) as any)[0];
                    const rest = input.slice(parsed.length);

                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: char + parsed.length,
                    });
                }
                return ok(false);
            }

            const tryParseWhiteSpace = isKeptWhiteSpace(documentPath, id);

            function tryParseWord(input: string, line: number, char: number): StepParseResult<string> {
                if(!opened) {
                    return ok('stop');
                }

                const parsed = input.charAt(0);
                const rest = input.slice(1);

                return ok({
                    type: 'parse result',
                    subResult: parsed,
                    rest,
                    line,
                    char: char + 1,
                });
            }

            const parser = parserBuilder.createParser(tryParseMultiline, tryParseWhiteSpace, tryParseWord);
            const parsed = parser.parse(toParse, startingLine, startingChar);

            if(parsed.success) {
                if(opened) {
                    return fail(`Multiline code block at { line: ${startingLine}, char: ${startingChar} } does not close`, documentPath);
                }
                const [peaces, leftover] = parsed.value;
                if(leftover.location.line === startingLine && leftover.location.char === startingChar) {
                    return ok(false);
                }

                const result = peaces.join('').trim();
                const rest = leftover.remaining;
                
                return ok({
                    type: 'parse result',
                    subResult: { location: { line: startingLine, char: startingChar }, type: 'text', text: result },
                    rest,
                    line: leftover.location.line,
                    char: leftover.location.char,
                });
            }

            return parsed;
        }
    }

    function isInline(documentPath: string): HandleValue<DocumentPart> {
        return function (toParse: string, startingLine: number, startingChar: number): StepParseResult<DocumentPart> {
            let opened: boolean = false;

            function tryParseInLine(input: string, line: number, char: number): StepParseResult<string> {
                if(doesIt.startWithInlineMarker.test(input)) {
                    opened = !opened;

                    const parsed: string = (input.match(doesIt.startWithInlineMarker) as any)[0];
                    const rest = input.slice(parsed.length);

                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: char + parsed.length,
                    });
                }
                return ok(false);
            }

            function tryParseWhiteSpace(input: string, line: number, char: number): StepParseResult<string> {
                if(doesIt.startWithWhiteSpace.test(input)) {
                    let doesItStartWithNewLine = /^\r|\n/;
                    if(doesItStartWithNewLine.test(input)) {
                        return fail(`Inline code block at { line: ${startingLine}, char: ${startingChar} } contains a new line before closing.`, documentPath);
                    }

                    const parsed: string = (input.match(doesIt.startWithWhiteSpace) as any)[0];
                    const rest = input.slice(parsed.length);
                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: char + parsed.length,
                    });
                }
                return ok(false);
            }

            function tryParseWord(input: string, line: number, char: number): StepParseResult<string> {
                if(!opened) {
                    return ok('stop');
                }

                let parsed = input.charAt(0);
                let rest = input.slice(1);
                return ok({
                    type: 'parse result',
                    subResult: parsed,
                    rest,
                    line,
                    char: char + 1,
                });
            }

            const parser = parserBuilder.createParser(tryParseInLine, tryParseWhiteSpace, tryParseWord);
            const parsed = parser.parse(toParse, startingLine, startingChar);

            if(parsed.success) {
                if(opened) {
                    return fail(`Inline code block at { line: ${startingLine}, char: ${startingChar} } does not close`, documentPath);
                }
    
                    const [parts, leftover] = parsed.value;
                if(leftover.location.line === startingLine && leftover.location.char === startingChar) {
                    return ok(false);
                }

                let result = parts.join('');

                return ok({
                    type: 'parse result',
                    subResult: { type: 'text', location: { line: startingLine, char: startingChar }, text: result },
                    rest: leftover.remaining,
                    line: leftover.location.line,
                    char: leftover.location.char,
                });
            }

            return parsed;
        }
    }

    function isDoculisp(documentPath: string, isOpen?: boolean): HandleValue<DocumentPart> {
        return function (toParse: string, startLine: number, startChar: number): StepParseResult<DocumentPart> {
            function updateChar(char: number, found: string): number {
                return char + found.length;
            }

            let depth = isOpen ? 1 : 0;
            isOpen = false;
            function tryParseDoculispOpen(input: string, line: number, char: number): StepParseResult<string> {
                if(doesIt.startWithDocuLisp.test(input)) {
                    if(0 < depth) {
                        return fail(`Doculisp Block at { line: ${startLine}, char: ${startChar} } contains an embedded doculisp block at { line: ${line}, char: ${char} }.`, documentPath);
                    }

                    const parsed: string = (input.match(doesIt.startWithDocuLisp) as any)[0];
                    const rest = input.slice(parsed.length);
                    depth++;

                    return ok({
                        type: 'discard',
                        rest,
                        line,
                        char: updateChar(char, parsed),
                    });
                }
                return ok(false);
            }

            function tryParseLispOpen(input: string, line: number, char: number): StepParseResult<string> {
                if(depth === 0) return ok(false);

                if(doesIt.startWithOpenLisp.test(input)) {
                    const parsed: string = (input.match(doesIt.startWithOpenLisp) as any)[0];
                    const rest = input.slice(parsed.length);
                    depth++;

                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: updateChar(char, parsed),
                    });
                }

                return ok(false);
            }

            function tryParseLispClose(input: string, line: number, char: number): StepParseResult<string> {
                if(depth === 0) return ok(false);
                
                if(doesIt.startsWithCloseLisp.test(input)) {
                    const parsed: string = (input.match(doesIt.startsWithCloseLisp) as any)[0];
                    const rest = input.slice(parsed.length);

                    const step: IParseStepForward = {
                        rest,
                        line,
                        char: updateChar(char, parsed),
                    };

                    depth--;

                    if(depth <= 0) {
                        return ok(parserBuilder.buildStepParse(step, {
                            type: 'discard',
                        }));
                    }

                    return ok(parserBuilder.buildStepParse(step, {
                        type: 'parse result',
                        subResult: parsed,
                    }));
                }
                return ok(false);
            }

            function tryParseWord(input: string, line: number, char: number): StepParseResult<string> {
                if(0 < depth){
                    const parsed = input.charAt(0);
                    const rest = input.slice(1);

                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: updateChar(char, parsed),
                    });
                }

                return ok('stop');
            }

            function tryParseWhiteSpace(input: string, line: number, char: number): StepParseResult<string> {
                if(0 < depth){
                    const tryParseWhiteSpace = isKeptWhiteSpace(documentPath, id);
                    return tryParseWhiteSpace(input, line, char);
                }
                return ok(false);
            }

            const parser = parserBuilder.createParser(tryParseDoculispOpen, tryParseLispOpen, tryParseLispClose, tryParseWhiteSpace, tryParseWord);
            const parsed = parser.parse(toParse, startLine, startChar);
            if(parsed.success) {
                if(0 < depth) {
                    return fail(`Doculisp block at { line: ${startLine}, char: ${startChar} } is not closed.`, documentPath);
                }

                const [parts, leftover] = parsed.value;
                if(leftover.location.line === startLine && leftover.location.char === startChar) {
                    return ok(false);
                }

                const step: IParseStepForward = {
                    rest: leftover.remaining,
                    line: leftover.location.line,
                    char: leftover.location.char,
                };

                if(0 === parts.length) {
                    return ok(parserBuilder.buildStepParse(step, {
                        type: 'discard'
                    }));
                }

                let result = parts.join('').trim();
                return ok(parserBuilder.buildStepParse(step, {
                    type: 'parse result',
                    subResult: { location: { line: startLine, char: startChar }, type: 'lisp', text: result },
                }));
            }
            return parsed;
        }
    }

    function isComment(documentPath: string): HandleValue<DocumentPart> {
        return function isComment(toParse: string, startingLine: number, startingChar: number): StepParseResult<DocumentPart> {
            let opened = false;
            const tryDoculisp = isDoculisp(documentPath);

            const stripWhiteSpace = isDiscardedWhiteSpace(doesIt, parserBuilder.createParser);

            function tryParseOpenComment(input: string, line: number, char: number): StepParseResult<DocumentPart> {
                if(doesIt.startWithOpenComment.test(input)) {
                    const parsed: string = (input.match(doesIt.startWithOpenComment) as any)[0];
                    opened = true;
                    return ok({
                        type: 'discard',
                        rest: input.slice(parsed.length),
                        line,
                        char: char + parsed.length,
                    });
                }
                
                return ok(false);
            }

            function tryParseCloseComment(input: string, line: number, char: number): StepParseResult<DocumentPart> {
                if(doesIt.startWithCloseComment.test(input)) {
                    if(opened){
                        const parsed: string = (input.match(doesIt.startWithCloseComment) as any)[0];
                        opened = false;

                        return ok({
                            type: 'discard',
                            rest: input.slice(parsed.length),
                            line,
                            char: char + parsed.length,
                        });
                    }

                    return ok('stop');
                }
                return ok(false);
            }

            function tryParseWhiteSpace(input: string, line: number, char: number): StepParseResult<DocumentPart> {
                if(opened && doesIt.startWithWhiteSpace.test(input)) {
                    return stripWhiteSpace(input, line, char);
                }
                return ok(false);
            }

            function tryParseDoculisp(input: string, line: number, char: number): StepParseResult<DocumentPart> {
                if(!opened) {
                    return ok(false);
                }

                return tryDoculisp(input, line, char);
            }

            function tryEndAll(input: string, line: number, char: number): StepParseResult<DocumentPart> {
                if(!opened) {
                    return ok('stop');
                }

                return ok({
                    type: 'discard',
                    rest: input.slice(1),
                    line,
                    char: char + 1,
                })
            }

            const parser = parserBuilder.createParser(tryParseOpenComment, tryParseCloseComment, tryParseWhiteSpace, tryParseDoculisp, tryEndAll);
            const parsed = parser.parse(toParse, startingLine, startingChar);
            if(parsed.success) {
                if(opened) {
                    return fail(`Open HTML Comment at { line: ${startingLine}, char: ${startingChar} } but does not close.`, documentPath);
                }

                const [result, leftover] = parsed.value;
                if(0 < result.length) {
                    return ok({
                        type: 'parse group result',
                        subResult: result.map(r => { return { type: 'keep', keptValue: r }}),
                        rest: leftover.remaining,
                        line: leftover.location.line,
                        char: leftover.location.char,
                    });
                }
                else if(leftover.location.line !== startingLine || leftover.location.char !== startingChar) {
                    return ok({
                        type: 'discard',
                        rest: leftover.remaining,
                        line: leftover.location.line,
                        char: leftover.location.char,
                    });
                }

                return ok(false);
            }

            return parsed;
        }
    }

    function isWord(documentPath: string) {
        return function isWord(toParse: string, startingLine: number, startingChar: number): StepParseResult<DocumentPart> {
            function tryParseEndParse(input: string, _line: number, _char: number): StepParseResult<string> {
                if(doesIt.startWithOpenComment.test(input)) {
                    return ok('stop')
                }
                if(doesIt.startWithInlineMarker.test(input)) {
                    return ok('stop')
                }
                return ok(false);
            }

            const tryParseWhiteSpace = isKeptWhiteSpace(documentPath, id);

            function tryParseWord(input: string, line: number, char: number): StepParseResult<string> {
                const startsWithWord = /^\S/;
                if(startsWithWord.test(input)) {
                    const parsed = input.charAt(0);
                    const rest = input.slice(1);
                    return ok({
                        type: 'parse result',
                        subResult: parsed,
                        rest,
                        line,
                        char: char + 1,
                    });
                }
                return ok(false);
            }

            const parser = parserBuilder.createParser(tryParseEndParse, tryParseWhiteSpace, tryParseWord);
            const parsed = parser.parse(toParse, startingLine, startingChar);

            if(parsed.success) {
                const [parts, leftover] = parsed.value;
                if(leftover.location.line === startingLine && leftover.location.char === startingChar) {
                    return ok(false);
                }

                const result = parts.join('').trim();
                return ok({
                    type: 'parse result',
                    subResult: { type: 'text', text: result, location: { line: startingLine, char: startingChar } },
                    rest: leftover.remaining,
                    line: leftover.location.line,
                    char: leftover.location.char,
                });
            }

            return parsed;
        }
    }

    function parse(documentText: string, documentPath: string): Result<DocumentMap> {
        const isDoculispFile = path.extname(documentPath) === '.dlisp';
        const toParse: string =
            isDoculispFile ?
            `${documentText})` :
            documentText;

        const parser = 
            isDoculispFile ?
            parserBuilder.createParser(isDoculisp(documentPath, true)) :
            parserBuilder.createParser(isDiscardedWhiteSpace(doesIt, parserBuilder.createParser), isMultiline(documentPath), isInline(documentPath), isComment(documentPath), isWord(documentPath));

        const parsed = parser.parse(toParse, 1, 1);

        if(parsed.success) {
            const [parts, leftover] = parsed.value;
            if(isDoculispFile && 0 < leftover.remaining.length) {
                return fail(`Doculisp block at { line: 1, char: 1 } has something not contained in parenthesis at { line: ${leftover.location.line}, char: ${leftover.location.char - 1} }.`, documentPath);
            }

            return ok(createMap(documentPath, parts));
        } else {
            return parsed;
        }
    }

    return parse;
}

const registerable: IRegisterable = {
    builder: (searches: Searcher, createParser: IInternals) => documentParse(searches.searchDocumentFor, createParser),
    name: 'documentParse',
    singleton: true,
    dependencies: ['searches', 'parser']
};

export {
    registerable as document,
};