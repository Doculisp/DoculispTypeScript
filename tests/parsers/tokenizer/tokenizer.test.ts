import { container } from "../../../src/container";
import { configure } from "approvals/lib/config";
import { getVerifier } from "../../tools";
import { Options } from "approvals/lib/Core/Options";
import { ITestableContainer } from "../../../src/types.containers";
import { TokenFunction } from '../../../src/types.tokens';
import { ILocation, Result, fail, ok } from "../../../src/types.general";
import { DocumentMap } from "../../../src/types.document";

describe('tokenizer', () => {
    let environment: ITestableContainer = undefined as any;
    let tokenizer: TokenFunction = undefined as any;
    let verifyAsJson: (data: any, options?: Options) => void = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        environment = container.buildTestable();
        tokenizer = environment.buildAs<TokenFunction>('tokenizer');
    });

    test('should fail if document parsing failed', () => {
        const parseResult = fail('This document did not parse', 'X:/non-exist.dlisp') as Result<DocumentMap>;

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    test('should return empty if given an empty parse result', () => {
        const parseResult: Result<DocumentMap> = ok({
            documentPath: 'c:/empty/readme.md',
            parts: [],
        });

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    test('should tokenize text as text', () => {
        const parseResult: Result<DocumentMap> = ok({
            documentPath: 'myText.md',
            parts: [
                {
                    type: 'text',
                    text: 'hello my text',
                    location: {
                        line: 5,
                        char: 23
                    },
                }
            ],
        });

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    describe('handling Doculisp', () => {
        test('should tokenize an empty comment', () => {
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(*)',
                        location: { line: 2, char: 1 },
                    },
                ],
            });
            
            let result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom', () => {
            const start: ILocation = { line: 4, char: 2 };
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with space after atom', () => {
            const start: ILocation = { line: 4, char: 2 };
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom )',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with new line after atom', () => {
            const start: ILocation = { line: 4, char: 2 };
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom\r\n)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom containing only numbers', () => {
            const start: ILocation = { line: 4, char: 2 };
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(123987)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with hyphen and underscore', () => {
            const start: ILocation = { line: 4, char: 2 };
            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'D:/comments/simple.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom-start_end)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should tokenize a single atom with a single word parameter', () => {
            const start: ILocation = { line: 1, char: 13 };

            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'Z:/parameter.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(the thing)',
                        location: start,

                    }
                ],
            }); 

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should tokenize a single atom with a multi word parameter', () => {
            const start: ILocation = { line: 1, char: 13 };

            let parseResult: Result<DocumentMap> = ok({
                documentPath: 'Z:/parameter.md',
                parts: [
                    {
                        type: 'lisp',
                        text: '(title the thing from beyond\n\tthe swamp)',
                        location: start,

                    }
                ],
            }); 

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
    });
});