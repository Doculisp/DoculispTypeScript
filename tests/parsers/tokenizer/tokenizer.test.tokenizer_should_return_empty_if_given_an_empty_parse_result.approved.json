{
  "success": true,
  "value": {
    "projectLocation": {
      "documentDepth": 4,
      "documentIndex": 8,
      "documentPath": "c:/empty/readme.md"
    },
    "tokens": []
  }
}
